### PR TITLE
Add support for empty enclosing types.

### LINKED ISSUE #570: Produce empty enclosing types when not explicitly marked.
Currently enclosing types are implicitly kept when a descendent is needed which also transitively keeps types of its members.

An example:
```proto
message FoodAndDrink {
  enum Bar { BEER_HALL }
  enum Food { MC_DONALDS }

  optional Food food = 1;
  optional Bar drink = 2;
}
message Person {
  optional FoodAndDrink.Bar drinking_at = 1;
}
```
When instructed to include only `Person` the generated code will contain `Food` as a result of `Bar`'s enclosing type `FoodAndDrink`'s member dependency.

Since `FoodAndDrink` is only needed for structure and not actually as a proto I would like it to become an empty type and its members not contribute dependencies. The generated code would instead look like this:
```java
public final class FoodAndDrink {
  private FoodAndDrink() {
    throw new AssertionError();
  }

  public enum Bar {
    BEER_HALL
  }
}
```

### PR DESCRIPTION
**Pull request by JakeWharton**
Closes #570.

### CODE REVIEW

File: wire-maven-plugin/src/main/java/com/squareup/wire/mojo/WireGenerateSourcesMojo.java [hunk 1]
```diff
@@ -95,10 +93,8 @@
 
         for (Type type : protoFile.types()) {
           Stopwatch stopwatch = Stopwatch.createStarted();
+          TypeSpec typeSpec = javaGenerator.generateType(type);
           ClassName javaTypeName = (ClassName) javaGenerator.typeName(type.type());
-          TypeSpec typeSpec = type instanceof MessageType
```
Comment (swankjesse) [inline wire-maven-plugin/src/main/java/com/squareup/wire/mojo/WireGenerateSourcesMojo.java]: Wow, we had this a lot more than I realized

File: wire-schema/src/main/java/com/squareup/wire/schema/EnclosingType.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  @Override public Location location() {
+    return location;
+  }
+
+  @Override public ProtoType type() {
+    return type;
+  }
+
+  @Override public String documentation() {
+    return documentation;
```
Comment (swankjesse) [inline wire-schema/src/main/java/com/squareup/wire/schema/EnclosingType.java]: Should this be used?

File: wire-schema/src/main/java/com/squareup/wire/schema/EnclosingType.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  }
+
+  /** Always null. */
```
Comment (swankjesse) [inline wire-schema/src/main/java/com/squareup/wire/schema/EnclosingType.java]: Not empty?

File: wire-schema/src/main/java/com/squareup/wire/schema/MessageType.java [hunk 4]
```diff
@@ -229,10 +229,14 @@ void validate(Linker linker) {
...
-    // If this type is not retained, and none of its nested types are retained, prune it.
     ImmutableList<Type> retainedNestedTypes = retainedNestedTypesBuilder.build();
-    if (!markSet.contains(protoType) && retainedNestedTypes.isEmpty()) {
-      return null;
+    if (!markSet.contains(protoType)) {
+      // If this type is not retained, and none of its nested types are retained, prune it.
+      if (retainedNestedTypes.isEmpty()) {
+        return null;
+      }
+      // If this type is not retained but retained nested types, replace it with an enclosing type.
```
Comment (swankjesse) [inline wire-schema/src/main/java/com/squareup/wire/schema/MessageType.java]: Dig.

**Comment by swankjesse** [comment]
LGTM! Nice model.

File: wire-schema/src/main/java/com/squareup/wire/schema/EnclosingType.java (hunk 3 above)
Comment (swankjesse) [inline wire-schema/src/main/java/com/squareup/wire/schema/EnclosingType.java]: Otherwise throw an unsupported operation exception?

Comment (JakeWharton) [inline wire-schema/src/main/java/com/squareup/wire/schema/EnclosingType.java]: Will throw

File: wire-schema/src/main/java/com/squareup/wire/schema/EnclosingType.java (hunk 2 above)
Comment (JakeWharton) [inline wire-schema/src/main/java/com/squareup/wire/schema/EnclosingType.java]: I had it originally (which is why the field and returning here exists) but removed it. I can add it back. I didn't have a strong opinion.

Comment (JakeWharton) [inline wire-schema/src/main/java/com/squareup/wire/schema/EnclosingType.java]: I updated to include it along with the warning.

**Pull request merged**